test/integration/listeners/docker/docker_listener_test.go

```diff
@@ -0,0 +1,189 @@
+	services := make(map[string]listeners.Service)
+	timeoutTicker := time.NewTicker(timeout)
+
+	for {
+		select {
+		case svc := <-channel:
+			for _, id := range containerIDs {
+				if string(svc.GetID()) == id {
+					services[id] = svc
+					if len(services) == len(containerIDs) {
+						return services, nil
+					}
+				} else {
+					log.Infof("%q != %q", svc.GetID(), id)
+
+				}
+			}
+			log.Infof("ignoring service from container ID %s", svc.GetID())
+		case <-timeoutTicker.C:
+			return services, fmt.Errorf("timeout listening for services, only got %d, expecting %d", len(services), len(containerIDs))
+		}
+	}
+}
+
+// Starts the listener AFTER the containers have started
+func (suite *DockerListenerTestSuite) TestListenAfterStart() {
+	suite.m.RLock()
+	defer suite.m.RUnlock()
+
+	containerIDs, err := suite.startContainers()
+	assert.Nil(suite.T(), err)
+	assert.Len(suite.T(), containerIDs, 2)
+	log.Infof("got container IDs %s from compose", containerIDs)
+
+	// Start listening after the containers started, they'll be listed in the init
+	suite.listener.Listen(suite.newSvc, suite.delSvc)
+
+	suite.commonSection(containerIDs)
+}
+
+// Starts the listener AFTER the containers have started
+func (suite *DockerListenerTestSuite) TestListenBeforeStart() {
+	suite.m.RLock()
+	defer suite.m.RUnlock()
+
+	// Start listening after the containers started, they'll be detected via docker events
+	suite.listener.Listen(suite.newSvc, suite.delSvc)
+
+	containerIDs, err := suite.startContainers()
+	assert.Nil(suite.T(), err)
+	assert.Len(suite.T(), containerIDs, 2)
+	log.Infof("got container IDs %s from compose", containerIDs)
+
+	suite.commonSection(containerIDs)
+}
+
+// Common section for both scenarios
+func (suite *DockerListenerTestSuite) commonSection(containerIDs []string) {
+	// We should get 2 new services
+	services, err := suite.getServices(containerIDs, suite.newSvc, 5*time.Second)
+	assert.Nil(suite.T(), err)
+	assert.Len(suite.T(), services, 2)
+
+	expectedIDs := make(map[string][]string)
+	for _, container := range containerIDs {
+		inspect, err := suite.dockerutil.Inspect(container, false)
+		assert.Nil(suite.T(), err)
+		if strings.Contains(inspect.Name, "redis-with-id") {
+			expectedIDs[container] = []string{"custom-id"}
+		} else {
+			expectedIDs[container] = []string{"docker://" + container, "redis"}
+		}
+	}
+
+	for _, service := range services {
+		pid, err := service.GetPid()
+		assert.Nil(suite.T(), err)
+		assert.True(suite.T(), pid > 0)
+		hosts, err := service.GetHosts()
+		assert.Nil(suite.T(), err)
+		assert.Len(suite.T(), hosts, 1)
+		ports, err := service.GetPorts()
+		assert.Nil(suite.T(), err)
+		assert.Len(suite.T(), ports, 1)
+		tags, err := service.GetTags()
+		assert.Nil(suite.T(), err)
+
+		assert.Contains(suite.T(), tags, "docker_image:redis:latest", "image_name:redis", "image_tag:latest")
```

Comment:
The signature of the function is `func Contains(t TestingT, s, contains interface{}, msgAndArgs ...interface{}) bool`
So in this case it should be replaced by 3 asserts.